mac/Keyman4MacIM/Keyman4MacIM/KMInputController.m

```diff
@@ -67,35 +58,74 @@ - (void)handleBackspace:(NSEvent *)event {
 }
 
 /**
- * The Keyman input method is deactivating because the user chose a different input method: notification from KMInputMethodLifecycle
+ * Called by the app delegate when KMInputMethodLifecycle determines the user has switched clients
  */
-- (void)inputMethodDeactivated:(NSNotification *)notification {
-  os_log_debug([KMLogs lifecycleLog], "***KMInputController inputMethodDeactivated, deactivating eventHandler");
-  if (_eventHandler != nil) {
-    [_eventHandler deactivate];
-  }
+- (void)changeClients:(NSString *)clientAppId {
+  os_log_debug([KMLogs lifecycleLog], "***KMInputController changeClients, deactivate old eventHandler and activate new one");
+  
+  [self deactivateEventHandler];
+  [self activateEventHandler:clientAppId];
 }
 
 /**
- * The user has switched to a different text input client: notification from KMInputMethodLifecycle
+ * Called by the app delegate when KMInputMethodLifecycle determines the user has deactivated Keyman
  */
-- (void)inputMethodChangedClient:(NSNotification *)notification {
-  os_log_debug([KMLogs lifecycleLog], "***KMInputController inputMethodChangedClient, deactivating old eventHandler and activating new one");
+- (void)deactivateClient {
+  os_log_debug([KMLogs lifecycleLog], "***KMInputController deactivateClient, deactivate old eventHandler");
+  
+  [self deactivateEventHandler];
+}
+
+- (void)deactivateEventHandler {
   if (_eventHandler != nil) {
+    os_log_debug([KMLogs lifecycleLog], " * KMInputController deactivate old eventHandler");
     [_eventHandler deactivate];
   }
-  _eventHandler = [[KMInputMethodEventHandler alloc] initWithClient:[KMInputMethodLifecycle getRunningApplicationId] client:self.client];
+}
 
+- (void)activateEventHandler:(NSString *)clientAppId {
+  os_log_debug([KMLogs lifecycleLog], " * KMInputController activate new eventHandler");
+  _eventHandler = [[KMInputMethodEventHandler alloc] initWithClient:clientAppId client:self.client];
 }
 
+/**
+ * Called by the OS to activate the input method, but some calls are not useful and are followed milliseconds later
+ * by a deactivateServer. These short-lived activations may be generated by clicking and releasing menu items without changing applications.
+ * Rather than treat every message received as a true activation, we send a message to KMInputMethodLifeCycle to evaluate.
+ */
 - (void)activateServer:(id)sender {
+  os_log_info([KMLogs lifecycleLog], " +++KMInputController, activateServer, self=%p", self);
   [sender overrideKeyboardWithKeyboardNamed:@"com.apple.keylayout.US"];
+  
+  /**
+   * When this KMInputController becomes the active server for the input method, then immediately update the AppDelegate.
+   * The duration that this controller is the active server may be extremely short, but, if so, we will receive another
+   * call to activateServer moments later and can update the AppDelegate again.
+   */
```

Comment:
```suggestion
  /*
   When this KMInputController becomes the active server for the input method, then immediately update the AppDelegate.
   The duration that this controller is the active server may be extremely short, but, if so, we will receive another
   call to activateServer moments later and can update the AppDelegate again.
   */
```

Note, inline comments should not have `/**` because that signals a javadoc style comment.

Reply:
So when can we use that style of comment? I thought you said not to use those for the header but that it was ok elsewhere. And what problem does that cause?

I can stop using those comments altogether, but I just think they are easier to read and contrast with the fixed-width text. Maybe I can customize the formatting in Xcode to get the same effect.

Reply:
Doxygen-style (Javadoc-style) comments should be used for function and class headers and the like. Comments mid-function should just use the language standard comment style, whatever that may be.

https://github.com/keymanapp/keyman/wiki/Keyman-Code-Style-Guide#Source-Code-Documentation-Syntax

I've made a tweak to the wiki to clarify.